File: firmware/src/decision.cpp
#include "decision.h"
#include <cstdio>
#include "pico/stdlib.h"
#include "../include/config.h"

DecisionEngine::DecisionEngine() 
    : activated(false), consecutive_hits(0), cooldown_until_ms(0) {}

DecisionEngine::~DecisionEngine() {}

bool DecisionEngine::begin() {
    activated = false;
    consecutive_hits = 0;
    cooldown_until_ms = 0;

    // Initialize LED and Buzzer pins as outputs
    gpio_init(PIN_LED_GREEN);
    gpio_set_dir(PIN_LED_GREEN, GPIO_OUT);
    gpio_init(PIN_LED_RED);
    gpio_set_dir(PIN_LED_RED, GPIO_OUT);
    gpio_init(PIN_BUZZER);
    gpio_set_dir(PIN_BUZZER, GPIO_OUT);

    // Set initial state: Red LED on (listening), Green LED and buzzer off
    gpio_put(PIN_LED_GREEN, 0);
    gpio_put(PIN_LED_RED, 1);
    gpio_put(PIN_BUZZER, 0);

    return true;
}

void DecisionEngine::processPrediction(int pred_class, float target_conf, uint32_t current_time_ms, bool vad_speech) {
    // 1. Cooldown Check
    if (current_time_ms > 0 && current_time_ms < cooldown_until_ms) {
        consecutive_hits = 0;
        activated = false;
        std::printf("[COOLDOWN] Ignoring wake-word detections (Cooldown active until %u ms)\n", cooldown_until_ms);
        updateActuators(pred_class, target_conf);
        return;
    }

    // 2. VAD Speech + Target Class + Threshold Check
    bool valid_hit = vad_speech && (pred_class == CLASS_ID_TARGET) && (target_conf >= WAKE_WORD_THRESHOLD);

    if (valid_hit) {
        consecutive_hits++;
        std::printf("[CONFIRMATION] Wake confirmation: %d/%d (Conf: %.2f%%)\n", 
                    consecutive_hits, CONFIRMATION_COUNT, target_conf * 100.0f);
        
        if (consecutive_hits >= CONFIRMATION_COUNT) {
            activated = true;
            if (current_time_ms > 0) {
                cooldown_until_ms = current_time_ms + COOLDOWN_MS;
            }
            consecutive_hits = 0;
            std::printf("[ACTIVATION] WAKE CONFIRMED! Triggering action. 1500 ms Cooldown active.\n");
        } else {
            activated = false;
        }
    } else {
        // Interruption: confidence below threshold or non-speech -> Reset counter
        consecutive_hits = 0;
        activated = false;
    }

    updateActuators(pred_class, target_conf);
}

void DecisionEngine::updateActuators(int pred_class, float target_conf) {
    if (activated) {
        std::printf("[ACTIVATION] WAKE WORD DETECTED! Class: Hero Arise | Conf: %.2f%%\n", target_conf * 100.0f);
        gpio_put(PIN_LED_GREEN, 1);
        gpio_put(PIN_LED_RED, 0);
        gpio_put(PIN_BUZZER, 1);
    } else if (consecutive_hits > 0) {
        std::printf("[CONFIRMING] Wake confirmation: %d/%d\n", consecutive_hits, CONFIRMATION_COUNT);
        gpio_put(PIN_LED_GREEN, 0);
        gpio_put(PIN_LED_RED, 1);
        gpio_put(PIN_BUZZER, 0);
    } else {
        std::printf("[IDLE] Listening... Class: %s | Target Conf: %.2f%%\n", 
                    (pred_class == CLASS_ID_UNKNOWN ? "Unknown" : "Noise"), 
                    target_conf * 100.0f);
        gpio_put(PIN_LED_GREEN, 0);
        gpio_put(PIN_LED_RED, 1);
        gpio_put(PIN_BUZZER, 0);
    }
}

bool DecisionEngine::isActivated() const {
    return activated;
}

int DecisionEngine::getConsecutiveHits() const {
    return consecutive_hits;
}

bool DecisionEngine::isInCooldown(uint32_t current_time_ms) const {
    return current_time_ms > 0 && current_time_ms < cooldown_until_ms;
}

File: ml/src/decision.py
import time
import sys
from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent.parent.parent
if str(BASE_DIR) not in sys.path:
    sys.path.insert(0, str(BASE_DIR))

# Existing threshold from config
WAKE_WORD_THRESHOLD = 0.75
CONFIRMATION_COUNT = 3
COOLDOWN_MS = 1500

class DecisionEngine:
    """
    Temporal Wake-Word Confirmation + Post-Trigger Cooldown Engine.
    
    Rules:
    1. Temporal Confirmation: Requires 3 consecutive inference frames with target confidence >= WAKE_WORD_THRESHOLD.
    2. Reset: Any frame dropping below threshold or non-speech VAD resets consecutive counter to 0 immediately.
    3. Cooldown: After successful trigger, ignore new wake-word detections for 1500 ms.
    """
    def __init__(self, threshold=WAKE_WORD_THRESHOLD, confirmation_count=CONFIRMATION_COUNT, cooldown_ms=COOLDOWN_MS):
        self.threshold = threshold
        self.confirmation_count = confirmation_count
        self.cooldown_ms = cooldown_ms
        
        self.consecutive_hits = 0
        self.cooldown_until_ms = 0
        self.last_status = "Listening"
        self.trigger_count = 0

    def reset(self):
        self.consecutive_hits = 0
        self.cooldown_until_ms = 0
        self.last_status = "Listening"
        self.trigger_count = 0

    def process_prediction(self, pred_class, target_conf, current_time_ms=None, vad_speech=True):
        """
        Processes a single prediction frame.
        
        Parameters:
        - pred_class: 0 (Hero Arise / target_word), 1 (unknown_words), 2 (background_noise)
        - target_conf: float confidence [0.0 - 1.0] for target class
        - current_time_ms: int timestamp in ms (if None, uses current system time)
        - vad_speech: bool (True if VAD energy gate detects speech)
        
        Returns:
        dict containing:
        - triggered (bool): True on the frame wake trigger occurs
        - status (str): UI status string
        - consecutive_hits (int): current consecutive confirmation count
        - in_cooldown (bool): True if currently within post-trigger cooldown window
        - trigger_count (int): total triggers fired
        """
        if current_time_ms is None:
            current_time_ms = int(time.time() * 1000)

        # 1. Check if currently in cooldown
        if current_time_ms < self.cooldown_until_ms:
            self.consecutive_hits = 0
            self.last_status = "Cooldown"
            return {
                "triggered": False,
                "status": "Cooldown",
                "consecutive_hits": 0,
                "in_cooldown": True,
                "trigger_count": self.trigger_count
            }

        # 2. Check if valid target wake-word detection frame
        valid_hit = vad_speech and (pred_class == 0) and (target_conf >= self.threshold)

        if valid_hit:
            self.consecutive_hits += 1
            if self.consecutive_hits >= self.confirmation_count:
                # WAKE CONFIRMED -> Fire trigger action and set 1500 ms cooldown
                self.trigger_count += 1
                self.cooldown_until_ms = current_time_ms + self.cooldown_ms
                self.consecutive_hits = 0
                self.last_status = "Wake confirmed"
                return {
                    "triggered": True,
                    "status": "Wake confirmed",
                    "consecutive_hits": self.confirmation_count,
                    "in_cooldown": True,
                    "trigger_count": self.trigger_count
                }
            else:
                self.last_status = f"Wake confirmation: {self.consecutive_hits}/{self.confirmation_count}"
                return {
                    "triggered": False,
                    "status": self.last_status,
                    "consecutive_hits": self.consecutive_hits,
                    "in_cooldown": False,
                    "trigger_count": self.trigger_count
                }
        else:
            # Interruption: Confidence dropped below threshold or non-speech -> Reset counter
            self.consecutive_hits = 0
            self.last_status = "Listening"
            return {
                "triggered": False,
                "status": "Listening",
                "consecutive_hits": 0,
                "in_cooldown": False,
                "trigger_count": self.trigger_count
            }

    def get_live_metrics(self, res, target_conf, vad_speech, latency_ms=0.0):
        """
        Extracts structured live telemetry metrics from decision engine state.
        """
        if res["status"] == "Wake confirmed":
            macro_state = "Wake confirmed"
            conf_display = f"{self.confirmation_count} / {self.confirmation_count}"
        elif res["in_cooldown"]:
            macro_state = "Cooldown"
            conf_display = f"0 / {self.confirmation_count}"
        elif res["consecutive_hits"] > 0:
            macro_state = "Speech detected"
            conf_display = f"{res['consecutive_hits']} / {self.confirmation_count}"
        elif vad_speech:
            macro_state = "Speech detected"
            conf_display = f"0 / {self.confirmation_count}"
        else:
            macro_state = "Listening"
            conf_display = f"0 / {self.confirmation_count}"

        vad_status = "Speech detected" if vad_speech else "Silence"

        return {
            "state": macro_state,
            "status": res["status"],
            "confidence": float(target_conf),
            "confidence_pct": float(target_conf * 100.0),
            "latency_ms": float(latency_ms),
            "confirmation": conf_display,
            "consecutive_hits": res["consecutive_hits"],
            "confirmation_count": self.confirmation_count,
            "vad_speech": bool(vad_speech),
            "vad_status": vad_status,
            "in_cooldown": bool(res["in_cooldown"]),
            "triggered": bool(res["triggered"]),
            "input_source": "Software/Test Audio",
            "runtime_environment": "PC"
        }

    def format_live_monitor(self, res, target_conf, vad_speech, latency_ms=0.0):
        """
        Formats a clean, lightweight terminal performance monitor panel string.
        """
        metrics = self.get_live_metrics(res, target_conf, vad_speech, latency_ms)
        cooldown_str = "YES (active)" if metrics["in_cooldown"] else "NO"
        
        lines = [
            "--------------------------------------------------",
            "             HERO ARISE — LIVE MONITOR            ",
            "--------------------------------------------------",
            f"STATE          : {metrics['state']}",
            f"CONFIDENCE     : {metrics['confidence_pct']:.1f} %",
            f"LATENCY        : {metrics['latency_ms']:.2f} ms (PC software)",
            f"CONFIRMATION   : {metrics['confirmation']}",
            f"VAD            : {metrics['vad_status']}",
            f"COOLDOWN       : {cooldown_str}",
            f"INPUT          : {metrics['input_source']}",
            f"RUNTIME        : {metrics['runtime_environment']}",
            "--------------------------------------------------"
        ]
        return "\n".join(lines)
